ipv8/peerdiscovery/network.py

```diff
@@ -183,13 +227,10 @@ def remove_by_address(self, address):
         """
         with self.graph_lock:
             self._all_addresses.pop(address, None)
-            to_remove = []
-            for i in range(len(self.verified_peers)):
-                if self.verified_peers[i].address == address:
-                    to_remove.insert(0, i)
-                    self.services_per_peer.pop(self.verified_peers[i].mid, None)
-            for index in to_remove:
-                self.verified_peers.pop(index)
+            # Note that the services_per_peer will never be 0, we abuse the lazy `or` to pop the peers from
+            # the services_per_peer mapping if they are no longer included. This is fast.
+            self.verified_peers = [peer for peer in self.verified_peers
+                                   if peer.address != address or self.services_per_peer.pop(peer.mid, None) == 0]
```

Comment:
`self.services_per_peer.pop(peer.mid, None) == 0` 
should this be comparison of the length instead?

Reply:
No, the reason for its existence is that it will never be `0`:

1. Because of the lazy-or this will only execute for peers which need to be deleted.
2. For each peer to be deleted we pop it out of the `services_per_peer`, but as the return value is never `0` and thus this part of the check is always `False`, it is not re-added/un-deleted (this always produces the same list as `[peer for peer in self.verified_peers if peer.address != address]`).
[2.b. We do it like this because Python can execute mappings faster than explicit loops.]